roachtest: run asyncpg test suite in nightlies

I ran the `asyncpg` test suite locally with the following commands:

```
git clone --recurse-submodules git@github.com:MagicStack/asyncpg.git
cd asyncpg
pip install -e .
# modified asyncpg/connect_utils.py to point at CRDB
python setup.py test
```

It passed except for two tests that check the Postgres port
```
======================================================================
FAIL: test_connect_params (test_connect.TestConnectParams)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rafiss/scratch/asyncpg/tests/test_connect.py", line 792, in test_connect_params
    self.run_testcase(testcase)
  File "/Users/rafiss/scratch/asyncpg/tests/test_connect.py", line 748, in run_testcase
    self.assertEqual(expected, result, 'Testcase: {}'.format(testcase))
AssertionError: Tuples differ: ([('host1', 5432), ('host2', 5432)], {'database': 'db', 'user': 'user'}) != ([('host1', 26257), ('host2', 26257)], {'user': 'user', 'database': 'db'})

First differing element 0:
[('host1', 5432), ('host2', 5432)]
[('host1', 26257), ('host2', 26257)]

- ([('host1', 5432), ('host2', 5432)], {'database': 'db', 'user': 'user'})
?              ^^^              ^^^

+ ([('host1', 26257), ('host2', 26257)], {'database': 'db', 'user': 'user'})
?             +++ ^             +++ ^
 : Testcase: {'name': 'dsn_only_multi_host', 'dsn': 'postgresql://user@host1,host2/db', 'result': ([('host1', 5432), ('host2', 5432)], {'database': 'db', 'user': 'user'})}

======================================================================
FAIL: test_connect_pgpass_regular (test_connect.TestConnectParams)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rafiss/scratch/asyncpg/tests/test_connect.py", line 845, in test_connect_pgpass_regular
    self.run_testcase({
  File "/Users/rafiss/scratch/asyncpg/tests/test_connect.py", line 748, in run_testcase
    self.assertEqual(expected, result, 'Testcase: {}'.format(testcase))
AssertionError: Tuples differ: ([('abc', 5432)], {'password': 'password from pgpas[46 chars]db'}) != ([('abc', 26257)], {'user': 'user', 'password': 'pa[47 chars]db'})

First differing element 0:
[('abc', 5432)]
[('abc', 26257)]

- ([('abc', 5432)],
?            ^^^

+ ([('abc', 26257)],
?           +++ ^

   {'database': 'db',
    'password': 'password from pgpass for user@abc',
    'user': 'user'}) : Testcase: {'dsn': 'postgres://user@abc/db?passfile=/var/folders/1_/_kcb28_94w16ml515w877ynm0000gn/T/tmpjh5aawll', 'result': ([('abc', 5432)], {'password': 'password from pgpass for user@abc', 'user': 'user', 'database': 'db'})}

----------------------------------------------------------------------
Ran 284 tests in 122.763s

FAILED (failures=2, skipped=3)
Test failed: <unittest.runner.TextTestResult run=284 errors=0 failures=2>
error: Test failed: <unittest.runner.TextTestResult run=284 errors=0 failures=2>
```

Let's turn this into a nightly test that automates all the above steps.

--- a/pkg/cmd/roachtest/cluster.go
+++ b/pkg/cmd/roachtest/cluster.go
@@ -1724,17 +1724,17 @@ func (c *clusterImpl) GitClone(
 	ctx context.Context, l *logger.Logger, src, dest, branch string, node option.NodeListOption,
 ) error {
 	return c.RunL(ctx, l, node, "bash", "-e", "-c", fmt.Sprintf(`'
-if ! test -d %s; then
-  git clone -b %s --depth 1 %s %s
+if ! test -d %[1]s; then
+  git clone -b %[2]s --depth 1 %[3]s %[1]s
 else
-  cd %s
+  cd %[1]s
   git fetch origin
-  git checkout origin/%s
+  git checkout origin/%[2]s
 fi
 '`, dest,
-		branch, src, dest,
-		dest,
-		branch))
+		branch,
+		src,
+	))
 }
 
 func roachprodArgs(opts []option.Option) []string {

--- a/pkg/cmd/roachtest/tests/BUILD.bazel
+++ b/pkg/cmd/roachtest/tests/BUILD.bazel
@@ -8,6 +8,7 @@ go_library(
         "activerecord_blocklist.go",
         "allocator.go",
         "alterpk.go",
+        "asyncpg.go",
         "autoupgrade.go",
         "backup.go",
         "build_info.go",

--- a/pkg/cmd/roachtest/tests/asyncpg.go
+++ b/pkg/cmd/roachtest/tests/asyncpg.go
@@ -1,0 +1,117 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package tests
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/spec"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+)
+
+func registerAsyncpg(r registry.Registry) {
+	runAsyncpg := func(
+		ctx context.Context,
+		t test.Test,
+		c cluster.Cluster,
+	) {
+		if c.IsLocal() {
+			t.Fatal("cannot be run in local mode")
+		}
+		node := c.Node(1)
+		t.Status("setting up cockroach")
+		c.Put(ctx, t.Cockroach(), "./cockroach", c.All())
+		c.Start(ctx, c.All())
+
+		version, err := fetchCockroachVersion(ctx, c, node[0])
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if err := alterZoneConfigAndClusterSettings(ctx, version, c, node[0]); err != nil {
+			t.Fatal(err)
+		}
+
+		t.Status("cloning asyncpg and installing prerequisites")
+
+		if err := gitCloneWithRecurseSubmodules(
+			ctx,
+			c,
+			t.L(),
+			"https://github.com/MagicStack/asyncpg.git",
+			"/mnt/data1/asyncpg",
+			"master",
+			node,
+		); err != nil {
+			t.Fatal(err)
+		}
+
+		if err := repeatRunE(
+			ctx,
+			t,
+			c,
+			node,
+			"install python and pip",
+			`sudo apt-get -qq install python3.7 python3-pip libpq-dev python-dev`,
+		); err != nil {
+			t.Fatal(err)
+		}
+
+		// Install postgresql for asyncpg requires pg_ctl and postgres executable to setup.
+		// See asyncpg's code here:
+		// https://github.com/MagicStack/asyncpg/blob/383c711eb68bc6a042c121e1fddfde0cdefb8068/asyncpg/cluster.py#L407-L419
+		if err := repeatRunE(
+			ctx,
+			t,
+			c,
+			node,
+			"install postgresql",
+			`sudo apt-get -y install postgresql`,
+		); err != nil {
+			t.Fatal(err)
+		}
+
+		if err := repeatRunE(
+			ctx,
+			t,
+			c,
+			node,
+			"install asyncpg's dependencies",
+			"cd /mnt/data1/asyncpg && sudo pip3 install django && pip3 install -e ."); err != nil {
+			t.Fatal(err)
+		}
+
+		const cockroachPort = 26257
+		if err := repeatRunE(
+			ctx,
+			t,
+			c,
+			node,
+			"run setup.py test for asyncpg",
+			fmt.Sprintf("cd /mnt/data1/asyncpg && PGPORT=%d python3 setup.py test", cockroachPort),
+		); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	r.Add(registry.TestSpec{
+		Name:    "asyncpg",
+		Owner:   registry.OwnerSQLExperience,
+		Cluster: r.MakeClusterSpec(1, spec.CPU(16)),
+		Tags:    []string{`default`, `orm`},
+		Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
+			runAsyncpg(ctx, t, c)
+		},
+	})
+}

--- a/pkg/cmd/roachtest/tests/canary.go
+++ b/pkg/cmd/roachtest/tests/canary.go
@@ -22,6 +22,7 @@ import (
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/logger"
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option"
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
 	"github.com/cockroachdb/cockroach/pkg/util/retry"
@@ -290,4 +291,28 @@ func repeatGetLatestTag(
 		return releaseTags[len(releaseTags)-1].tag, nil
 	}
 	return "", fmt.Errorf("could not get tags from %s, due to error: %s", url, lastError)
+}
+
+// gitCloneWithRecurseSubmodules clones a git repo from src into dest and checks out origin's
+// version of the given branch, but with a --recurse-submodules flag.
+// The src, dest, and branch arguments must not contain shell special characters.
+func gitCloneWithRecurseSubmodules(
+	ctx context.Context,
+	c cluster.Cluster,
+	l *logger.Logger,
+	src, dest, branch string,
+	node option.NodeListOption,
+) error {
+	return c.RunL(ctx, l, node, "bash", "-e", "-c", fmt.Sprintf(`'
+if ! test -d %[1]s; then
+  git clone --recurse-submodules -b %[2]s --depth 1 %[3]s %[1]s
+else
+  cd %[1]s
+  git fetch origin
+  git checkout origin/%[2]s
+fi
+'`, dest,
+		branch,
+		src,
+	))
 }

--- a/pkg/cmd/roachtest/tests/registry.go
+++ b/pkg/cmd/roachtest/tests/registry.go
@@ -18,6 +18,7 @@ func RegisterTests(r registry.Registry) {
 	registerActiveRecord(r)
 	registerAllocator(r)
 	registerAlterPK(r)
+	registerAsyncpg(r)
 	registerAutoUpgrade(r)
 	registerBackup(r)
 	registerBackupNodeShutdown(r)

